backends/dpdk/dpdkHelpers.cpp

```diff
@@ -184,6 +184,75 @@ bool ConvertStatementToDpdk::preorder(const IR::AssignmentStatement *a) {
             } else {
                 BUG("%1% Not implemented", e->originalExternType->name);
             }
+        } else if (auto e = mi->to<P4::ExternFunction>()) {
+            /* PNA SelectByDirection extern is implemented as
+                SelectByDirection<T>(in PNA_Direction_t direction, in T n2h_value, in T h2n_value) {
+                    if (direction == PNA_Direction_t.NET_TO_HOST) {
+                        return n2h_value;
+                    } else {
+                        return h2n_value;
+                    }
+                }
+                Example:
+                    table ipv4_da_lpm {
+                        key = {
+                            SelectByDirection(istd.direction, hdr.ipv4.srcAddr, hdr.ipv4.dstAddr):
+                                                                          lpm @name ("ipv4_addr");
+                        }
+                        ....
+                }
+
+                In this example, KeySideEffect pass inserts a temporary for holding the result of
+                complex key expression and replaces the key expression with the temporary. An
+                assignment is inserted in the apply block before the table apply to assign the complex
+                expression into this temporary variable. After KeySideEffect pass, SelectByDirection
+                extern invocation is present as RHS of assignment and hence is evaluated here in this
+                visitor.
+
+                After keySideEffect pass:
+                bit <> key_0;
+                ...
+                table ipv4_da_lpm {
+                    key = {
+                        key_0: lpm @name ("ipv4_addr");
+                    }
+                    ....
+                }
+
+                apply {
+                    key_0 = SelectByDirection<bit<32>>(istd.direction, hdr.ipv4.srcAddr, hdr.ipv4.dstAddr);
+                    ...
+                }
+
+                This is replaced by an assignment of the resultant value into a temporary.
+                if (istd.direction == PNA_Direction_t.NET_TO_HOST) {
+                    key_0 = hdr.ipv4.srcAddr;
+                } else {
+                    key_0 = hdr.ipv4.dstAddr;
+                }
+            */
+            if (e->method->name == "SelectByDirection") {
+                auto args = e->expr->arguments;
+                if (args->size() != 3)
+                    ::error("Unexpected number of arguments\n");
```

Comment:
Please use the error function that specifies an error type. Also, you should show something about the source code, otherwise no one will understand what the problem is when it occurs. This is probably a modelError.

Reply:
These error checks may not be required here. Frontend/typechecker is already checking if the number of arguments for extern invocation match with the declaration or not.